crates/sui-bytecode-src-verifier/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt::{Debug, Display},
+    str::FromStr,
+};
+
+use move_compiler::compiled_unit::CompiledUnitEnum;
+use move_core_types::account_address::AccountAddress;
+use move_package::compilation::compiled_package::CompiledPackage;
+use move_symbol_pool::Symbol;
+
+use sui_sdk::{
+    rpc_types::{SuiRawData, SuiRawMoveObject, SuiRawMovePackage},
+    ReadApi,
+};
+use sui_types::{
+    base_types::{ObjectID, ObjectIDParseError},
+    error::SuiError,
+};
+
+#[derive(Clone, Debug)]
+pub struct DependencyVerificationResult {
+    pub verified_dependencies: HashMap<AccountAddress, Dependency>,
+}
+
+#[derive(Debug)]
+pub enum DependencyVerificationError {
+    /// Could not resolve Sui addresses for package dependencies
+    ResolutionGraphNotResolved(anyhow::Error),
+    /// Could not convert a dependencies' resolved Sui address to a Sui object ID
+    ObjectIdFromAddressFailure(ObjectIDParseError),
+    /// Could not read a dependencies' on-chain object
+    DependencyObjectReadFailure(anyhow::Error),
+    /// Dependency object does not exist or was deleted
+    SuiObjectRefFailure(SuiError),
+    /// Dependency address contains a Sui object, not a Move package
+    ObjectFoundWhenPackageExpected(ObjectID, SuiRawMoveObject),
+    /// A local dependency was not found
+    ///
+    /// params:  package, module
+    LocalDependencyNotFound(Symbol, Option<Symbol>),
+    /// An on-chain version of a local dependency was not found
+    ///
+    /// params:  package, module
+    OnChainDependencyNotFound(Symbol, Option<Symbol>),
+    /// Local dependencies have a different number of modules than on-chain
+    ///
+    /// params:  expected count, on-chain count, missing
+    ModuleCountMismatch(usize, usize, Vec<String>),
+    /// A local dependency module did not match its on-chain version
+    ///
+    /// params:  package, module, address
+    ModuleBytecodeMismatch(String, String, AccountAddress),
```

Comment:
nit: if the params need documenting, would named parameters work better?  Seems reasonable that if the param names are useful when reading the definitions, they would be useful when reading the uses as well.

```suggestion
    /// A local dependency was not found
    LocalDependencyNotFound { package: Symbol, module: Option<Symbol> },
    /// An on-chain version of a local dependency was not found
    OnChainDependencyNotFound { package: Symbol, module: Option<Symbol> },
    /// Local dependencies have a different number of modules than on-chain
    ModuleCountMismatch { expected: usize, on_chain: usize, missing: Vec<String> },
    /// A local dependency module did not match its on-chain version
    ModuleBytecodeMismatch { package: String, module: String, address: AccountAddress },
```